src/http2.rs

```diff
@@ -0,0 +1,563 @@
+                                       "Not in a writable state")),
+            Some(ref mut req) => req.body.flush(),
+        }
+    }
+}
+
+impl<S> Read for Http2Message<S> where S: CloneableStream {
+    #[inline]
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        match self.response {
+            None => Err(io::Error::new(io::ErrorKind::Other,
+                                       "Not in a readable state")),
+            Some(ref mut resp) => resp.body.read(buf),
+        }
+    }
+}
+
+impl<S> Http2Message<S> where S: CloneableStream {
+    /// A helper method that prepares the path of a request by extracting it from the given `Url`.
+    fn prepare_path(&mut self, url: Url) -> Vec<u8> {
+        let mut uri = url.serialize_path().unwrap();
+        if let Some(ref q) = url.query {
+            uri.push('?');
+            uri.push_str(&q[..]);
+        }
+        uri.into_bytes()
+    }
+
+    /// A helper method that prepares the headers that should be sent in an HTTP/2 request.
+    ///
+    /// Adapts the `Headers` into a list of octet string pairs.
+    fn prepare_headers(&mut self, mut headers: Headers) -> Vec<Http2Header> {
+        if headers.remove::<header::Connection>() {
+            warn!("The `Connection` header is not valid for an HTTP/2 connection.");
+        }
+        // HTTP/2 header names MUST be lowercase.
+        headers.iter().map(|h| {
+            (h.name().to_ascii_lowercase().into_bytes(), h.value_string().into_bytes())
+        }).collect()
+    }
+
+    /// A helper method that prepares the body for sending in an HTTP/2 request.
+    #[inline]
+    fn prepare_body(&mut self, body: Vec<u8>) -> Option<Vec<u8>> {
+        if body.len() == 0 {
+            None
+        } else {
+            Some(body)
+        }
+    }
+
+    /// Parses the response, as returned by `solicit`, into a `ResponseHead` and the full response
+    /// body.
+    ///
+    /// Returns them as a two-tuple.
+    fn parse_response(&mut self, response: ::solicit::http::Response) -> (ResponseHead, Vec<u8>) {
+        // - The status code
+        let status = response.status_code().unwrap();
+
+        // - The headers
+        let mut headers = Headers::new();
+        // Headers are moved into `hyper`'s `Headers` collection; pseudo-headers
+        // are ignored (status has already been extracted).
+        // TODO: Coalesce possible headers with the name key into a single vector.
+        //       `solicit` returns the raw original header list and lets the clients
+        //       figure out how they would like to transform it.
+        for (name, value) in response.headers.into_iter() {
+            // Unfortunately, `hyper` wants its header names to be strings...
+            // We just panic if it ends up being an invalid utf8 string, but no
+            // copies are made.
+            headers.set_raw(String::from_utf8(name).unwrap(), vec![value]);
```

Comment:
The names should still be valid utf8 ([8.1.2 HTTP Header Fields](http://http2.github.io/http2-spec/#HttpHeaders)) on the HTTP layer, that's true. (HPACK is what treats both the name and value as an opaque octet sequence.) So, indeed, if the name cannot be decoded as utf8, the response is malformed. This is now fixed (along with a test).

As for multiple `SetCookie` headers -- the situation is unchanged: multiple cookies correspond to multiple `set-cookies` headers. That said, I was hoping that `hyper` would correctly handle this itself...

What I see now, though, is that it effectively fakes it by having the *value* of the `Set-Cookie` entry in the `header::Headers` struct contain a CRLF delimiter and continue on to the next `Set-Cookie` header.

```rust
    let cookies = header::SetCookie(vec![
        cookie::Cookie::new("foo".to_owned(), "bar".to_owned()),
        cookie::Cookie::new("baz".to_owned(), "quux".to_owned())
    ]);
    // Now the following is true:
    cookies.name() == "Set-Cookie"
    cookies.value_string() == "foo=bar; Path=/\r\nSet-Cookie: baz=quux; Path=/"
```

That hack works when sending the header over HTTP/1.1, but not HTTP/2, as the CRLF is meaningless. I've worked around it in the `prepare_headers` function to special case the `Set-Cookie` and extract the cookies that it contains into separate HTTP/2 headers (although since for now there's only client support, it shouldn't actually get triggered, as clients aren't supposed to send that header).

On the other hand, when parsing a response with multiple `set-cookie` headers returned by an HTTP/2 server, it seems that `hyper`'s `Headers::from_raw` method ends up returning a headers collection with a `SetCookie` header that correctly contains all the cookies, as verified by a new unit test... Should `from_raw` not be relied on to do the right thing?